From 048f6b4b55d1ac777d76e1e5d496079e60cbda51 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Enrique=20Oca=C3=B1a=20Gonz=C3=A1lez?= <eocanha@igalia.com>
Date: Mon, 17 Jun 2019 17:00:56 +0000
Subject: [PATCH 7/9] Manual revert of 7b6be34 omxbufferpool: deallocate OMX
 buffers when stopping

Causes crashes when the app (eg: WebKit) still owns EGL buffers after the
pipeline has been destroyed. When the app releases those buffers and the last
one is returned to the pool, the pool is stopped and deallocates its buffers.
This triggers an gst_omx_port_update_port_definition() on port 221 (eglrender
out port), but the eglrenderer component might have already been destroyed
because the port has a dangling pointer to it which points to corrupt data.

This commit reverts the "deallocate when stopping" behaviour but keeps the
rest of the style changes of the original commit in order to avoid future
conflicts.

This commits also reverts part of 783e58fc "omxbufferpool: refactor to allow
memory sharing". That commit might have to be fully reverted if this causes
problems in the future.
---
 omx/gstomxbufferpool.c | 11 -----------
 omx/gstomxvideodec.c   |  4 ++--
 2 files changed, 2 insertions(+), 13 deletions(-)

diff --git a/omx/gstomxbufferpool.c b/omx/gstomxbufferpool.c
index b885685..303b122 100644
--- a/omx/gstomxbufferpool.c
+++ b/omx/gstomxbufferpool.c
@@ -186,17 +186,6 @@ gst_omx_buffer_pool_stop (GstBufferPool * bpool)
   /* Remove any buffers that are there */
   g_ptr_array_set_size (pool->buffers, 0);
 
-  GST_DEBUG_OBJECT (pool, "deactivating OMX allocator");
-  gst_omx_allocator_set_active (pool->allocator, FALSE);
-
-  /* ensure all memories have been deallocated;
-   * this may take a while if some memories are being shared
-   * and therefore are in use somewhere else in the pipeline */
-  gst_omx_allocator_wait_inactive (pool->allocator);
-
-  GST_DEBUG_OBJECT (pool, "deallocate OMX buffers");
-  gst_omx_port_deallocate_buffers (pool->port);
-
   if (pool->caps)
     gst_caps_unref (pool->caps);
   pool->caps = NULL;
diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index a349d1a..857371d 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -1212,6 +1212,8 @@ done:
 static gboolean
 gst_omx_video_dec_deallocate_output_buffers (GstOMXVideoDec * self)
 {
+  OMX_ERRORTYPE err;
+
   if (self->out_port_pool) {
     /* Pool will free buffers when stopping */
     gst_buffer_pool_set_active (self->out_port_pool, FALSE);
@@ -1221,8 +1223,6 @@ gst_omx_video_dec_deallocate_output_buffers (GstOMXVideoDec * self)
     GST_OMX_BUFFER_POOL (self->out_port_pool)->deactivated = TRUE;
     gst_object_unref (self->out_port_pool);
     self->out_port_pool = NULL;
-  } else {
-    OMX_ERRORTYPE err;
 
 #if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_GL)
     err =
-- 
2.34.1

